### PR TITLE
Add test for a failing migration from SLES to SLES4SAP and the successful rollback

### PR DESCRIPTION
**Pull request by bs-suse**
This test checks if a successful rollback happens on a failing migration from SLES to SLES4SAP in case that an invalid registration key is used.

- Verification run: http://keks.qa.suse.de/tests/206#step/migrate_sles_to_sles4sap/37
- Related ticket: https://jira.suse.com/browse/TEAM-2123

### CODE REVIEW

File: tests/sles4sap/migrate_sles_to_sles4sap.pm [hunk 1]
```diff
@@ -48,8 +52,16 @@ sub run {
     send_key 'ret';
     wait_serial "Please enter your activation code", timeout => 5;
     enter_cmd "${regcode}";
-    assert_script_run "ls /tmp/OK";
-    zypper_call "in -y -t pattern sap_server";
+
+    # test either a failing migration or a working one
+    if ("${regcode}" eq "invalid_key") {
```
Comment (ricardobranco777) [inline tests/sles4sap/migrate_sles_to_sles4sap.pm:57]: Perl is not Bash :)

I think just `$regcode` will work.

**Comment by bs-suse** [comment]
The test as it worked before the change is as well still working: http://keks.qa.suse.de/tests/203#step/migrate_sles_to_sles4sap/37

**Review by ricardobranco777 — COMMENTED** [review]
Don't know why we have Bashisms like `"${regcode} "`.

Otherwise LGTM

File: tests/sles4sap/migrate_sles_to_sles4sap.pm (hunk 1 above)
Comment (bs-suse) [reply tests/sles4sap/migrate_sles_to_sles4sap.pm:57]: Good point, I just wanted to be consistent with what was already there 3 lines above.

Comment (bs-suse) [reply tests/sles4sap/migrate_sles_to_sles4sap.pm]: I changed both occurrences of the bashism :-)

**Review by alvarocarvajald — COMMENTED** [review]
Some general comments:

- Due to the size and complexity of the changes, it will be better to squash all 4 commits into one.
- You can add in the description the Jira task as the related ticket. That way we have a reference as to where this test scenario was requested.

I have some comments below regarding the code, but it is looking good so far.

File: tests/sles4sap/migrate_sles_to_sles4sap.pm (hunk 1 above)
Comment (alvarocarvajald) [reply tests/sles4sap/migrate_sles_to_sles4sap.pm]: Sometimes "bashisms" are necessary:
```
# perl
$msg = 'Hello ';
print "Without bashisms: $msgWorld\n";
print "With    bashisms: ${msg}World\n";
Without bashisms: 
With    bashisms: Hello World
```
Or when doing some more complex data structures handling:
```
    for my $arg (@{$param_file}) {
        next unless $arg =~ $PARAM_RE;
        push @{$params{$1}}, path($2)->slurp;
    }
```
But agree with @ricardobranco777 .... it does look better without the curly braces. :)

**Review by ricardobranco777 — APPROVED** [review]

**Review by alvarocarvajald — APPROVED** [review]
LGTM

**Pull request merged**